query_optimizer/ExecutionGenerator.cpp

```diff
@@ -332,6 +337,63 @@ void ExecutionGenerator::createTemporaryCatalogRelation(
   insert_destination_proto->set_relation_id(output_rel_id);
 }
 
+void ExecutionGenerator::createPartitionedTemporaryCatalogRelation(
+    const P::PhysicalPtr &physical,
+    const CatalogRelation **catalog_relation_output,
+    S::InsertDestination *insert_destination_proto,
+    const attribute_id part_attr_id,
+    const std::size_t num_partitions) {
+  attribute_id pid;
+  std::unique_ptr<CatalogRelation> catalog_relation(
+      new CatalogRelation(optimizer_context_->catalog_database(),
+                          getNewRelationName(),
+                          -1 /* id */,
+                          true /* is_temporary*/));
+  attribute_id aid = 0;
+  for (const E::NamedExpressionPtr &project_expression :
+       physical->getOutputAttributes()) {
+    // The attribute name is simply set to the attribute id to make it distinct.
+    std::unique_ptr<CatalogAttribute> catalog_attribute(
+        new CatalogAttribute(catalog_relation.get(),
+                             std::to_string(aid),
+                             project_expression->getValueType(),
+                             aid,
+                             project_expression->attribute_alias()));
+    attribute_substitution_map_[project_expression->id()] =
+        catalog_attribute.get();
+    if (part_attr_id == catalog_attribute.get()->getID()) {
+      pid = aid;
+    }
+    catalog_relation->addAttribute(catalog_attribute.release());
+    ++aid;
+  }
+
+  PartitionSchemeHeader *part_scheme_header = new HashPartitionSchemeHeader(num_partitions, pid);
```

Comment:
Could we have a `TODO` comment stating that we would be able to choose the partition scheme based on the input relation, instead of using hash-based by default.